include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h

```diff
@@ -131,25 +131,44 @@ struct sycl_iterator
     }
 };
 
-// mode converter when property::noinit present
-template <access_mode __mode>
-struct _ModeConverter
+// map access_mode tag to access_mode value
+template <typename _ModTagT, typename _NoInitT = void>
+struct __access_mode_resolver
 {
-    static constexpr access_mode __value = __mode;
 };
 
 template <>
-struct _ModeConverter<access_mode::read_write>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::read_only)>, void>
 {
-    static constexpr access_mode __value = access_mode::discard_read_write;
+    static constexpr access_mode __mode = access_mode::read;
+};
+
+template <>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::write_only)>, void>
+{
+    static constexpr access_mode __value = access_mode::write;
+};
+
+template <>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::read_write)>, void>
+{
+    static constexpr access_mode __value = access_mode::read_write;
 };
 
+// map access_mode if property::noinit present
+// TODO: remove since discard_read_write and discard_write are deprecated in SYCL 2020
 template <>
-struct _ModeConverter<access_mode::write>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::write_only)>, __dpl_sycl::__no_init>
 {
     static constexpr access_mode __value = access_mode::discard_write;
 };
 
+template <>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::read_write)>, __dpl_sycl::__no_init>
+{
+    static constexpr access_mode __value = access_mode::discard_read_write;
```

Comment:
I suppose it's a matter of preference, but would it make sense to combine the init/no_init overloads? Something like:
```c++
template <typename _NoInitT>
struct __access_mode_resolver<std::decay_t<decltype(sycl::read_write)>, _NoInitT>
{
    static constexpr access_mode __mode = std::is_same_v<_NoInitT, void> ? access_mode::read_write : access_mode::discard_read_write;
};
```

Reply:
Thanks, I've applied it.

Reply:
Doesn't this kill the compile-time part of this?  
I thought the conditional operator `?` was not usable at compile time in a `constexpr`.

Reply:
Never mind, I think I have a case of Friday afternoon brain.  

I checked it out on godbolt and it seems to be fine in this context, as well as being used like this plenty of other places in the codebases.  I need to remind myself the details of `?` and `constexpr` because I know there is some nuance, but I'll withdraw my objection.